src/compiler/moduleNameResolver.ts

```diff
@@ -3332,8 +3332,8 @@ function resolveFromTypeRoot(moduleName: string, state: ModuleResolutionState) {
 // Program errors validate that `noEmit` or `emitDeclarationOnly` is also set,
 // so this function doesn't check them to avoid propagating errors.
 /** @internal */
-export function shouldAllowImportingTsExtension(compilerOptions: CompilerOptions, fromFileName?: string) {
-    return !!compilerOptions.allowImportingTsExtensions || fromFileName && isDeclarationFileName(fromFileName);
+export function shouldAllowImportingTsExtension(compilerOptions: CompilerOptions, fromFileName?: string): boolean | undefined {
+    return !!compilerOptions.allowImportingTsExtensions || (fromFileName !== undefined) && isDeclarationFileName(fromFileName);

```

Comment:
This is technically not semantically equivalent, no? Should be `!!fromFileName`?